src/librustc_mir/const_eval.rs

```diff
@@ -426,6 +427,16 @@ pub fn const_eval_provider<'a, 'tcx>(
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
 ) -> ::rustc::mir::interpret::ConstEvalResult<'tcx> {
+    if key.param_env.reveal == Reveal::All {
+        let mut key = key.clone();
+        key.param_env.reveal = Reveal::UserFacing;
```

Comment:
Please explain in a comment why we even try it with a different than the requested reveal first.

Reply:
Also, why do *both* queries need that logic? Shouldn't one be enough?

Reply:
if only `const_eval_raw` did it, we might duplicate validation even if the computation of the constant were deduplicated.

if only `const_eval` did it, we might duplicate the computation of the constant in the few explicit calls to `const_eval_raw`.